TESTS/mbed_drivers/sai/main.cpp

```diff
@@ -0,0 +1,293 @@
+
+            const uint32_t data_mask = ((1 << test_sai_formats[f_idx].data_length) - 1);
+
+            /* Set to unexpected. */
+            set_buffer(receive_buffer, sizeof(receive_buffer), 0xFF);
+
+            for (int i = 0; i < BUFFER_SIZE; i++) {
+                transmit_buffer[i] = data_mask / 100 * i;
+            }
+
+            uint32_t write_cnt = 0;
+            uint32_t read_cnt = 0;
+            uint32_t sample;
+
+            /* Send samples and read them back. */
+            while (1) {
+                if (write_cnt < BUFFER_SIZE) {
+                    sample = transmit_buffer[write_cnt];
+                    if (sai_transmitter.xfer(&sample) == true) {
+                        write_cnt++;
+                    }
+                }
+
+                if (sai_receiver.xfer(&sample) == true) {
+                    receive_buffer[read_cnt] = sample;
+                    read_cnt++;
+                }
+
+                if (read_cnt == BUFFER_SIZE) {
+                    break;
+                }
+            }
+
+            if (compare_buffers(transmit_buffer, receive_buffer, BUFFER_SIZE) == true) {
+                status = LOOPBACK_TEST_OK;
+            } else {
+                status = LOOPBACK_TEST_FAILURE;
+            }
+        } else {
+            status = LOOPBACK_TEST_SKIP;
+        }
+
+        /* First `required_formats_count` formats are obligate, so we expect that
+         * communication tests passes. Others formats may be supported, so test can pass or be skipped.
+         */
+        sprintf(message, "format idx: %lu", f_idx);
+        if (f_idx < required_formats_count) {
+            TEST_ASSERT_EQUAL_MESSAGE(LOOPBACK_TEST_OK, status, message);
+        } else {
+            TEST_ASSERT_NOT_EQUAL_MESSAGE(LOOPBACK_TEST_FAILURE, status, message);
+        }
+    }
+}
+
+/* Test that SAITransmitter class is able to send samples in loop-back mode.
+ * Test that SAIReceiver class is able to receive samples in loop-back mode.
+ *
+ * Given are SAITransmitter and SAIReceiver Classes and SAI loop-back connection.
+ * When SAI object is configured as transmitter with the specified format and
+ *      another SAI object is configured as receiver with the same format used.
+ * Then transmitter can successfully transmit samples to the receiver.
+ */
+void SAITransmitter_SAIReceiver_transmission_test()
+{
+    loopback_test_tatus_t status;
+
+    if (!LOOPBACK_CONNECTION) {
+        TEST_IGNORE_MESSAGE("NO LOOPBACK CONNECTION - TEST SKIPPED!");
+        return;
+    }
+
+    char message[50];
+
+    for (uint32_t f_idx = 0; f_idx < (sizeof(test_sai_formats) / sizeof(sai_format_t)); f_idx++) {
+
+        SAITransmitter sai_transmitter(SAI_A_MCLK, SAI_A_BCLK, SAI_A_WCLK, SAI_A_SD, &test_sai_formats[f_idx]);
+        SAIReceiver sai_receiver(SAI_B_MCLK, SAI_B_BCLK, SAI_B_WCLK, SAI_B_SD, &test_sai_formats[f_idx]);
+
+        // we should check init status before the actual test - for now there is no such option
```

Comment:
This is fine as is, just noting the following incoming PR to add the status check function:

https://github.com/ARMmbed/mbed-os/pull/6777

Reply:
I extend the test when status check class function will be added.